File: valid-parentheses/valid-parentheses.py
from typing import List 

"""
- use a stack
- add open brace/bracket opts to stack
- when encounter a closing brace/bracket pop() from stack
- verify

ex:
- s = '{[]}'
add '{' to stack
add '[' to stack 


"""


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 1: 
            return False 
        valid_keys = {
                      "}": "{", 
                      "]": "[",
                      ")": "("
                     }
        stack: List[str] = list() 
        for ch in s:
            if ch in valid_keys:
                if len(stack) > 0:
                    valid_open_brace = valid_keys[ch]
                    if not valid_open_brace == stack.pop():
                        return False 
                else:
                    return False 
            elif not ch in valid_keys: #append open brace types
                stack.append(ch)
            
        if len(stack) > 0: 
            return False 
        return True  

File: valid-parentheses/valid-parentheses.go
/*
time: O(n); len of string
space: O(n); stack can possibly be filled with all open bracket types
*/

func isValid(s string) bool {
    if len(s) == 1 {
        return false
    }
    charsMap := map[string]string{
        "}" : "{",
        "]" : "[",
        ")" : "(",
    }
    stack := []string{} 
    for _, v := range s {
        if openBracket, ok := charsMap[string(v)]; ok {
            if len(stack) == 0 {
                return false
            }
            lifoItem := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            if  openBracket != lifoItem {
                return false
            }
            continue
        }
        stack = append(stack, string(v))
       
    }
    if len(stack) > 0 {
        return false
    }
    return true 
    
}